EZP-25487: Stored Image Variation in Image Variation Cache

alongosz opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-25487](https://jira.ez.no/browse/EZP-25487)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `6.7`, `6.13`, `7.1`, `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes\*

## Summary
This is #2121 by @kmadejski including minor improvements.

This approach uses Persistence Cache to store Image variation. The only difference when compared to #2121 is type-hinting `Psr\Cache\` interface instead of the one from `Stash` to make merging up to 7.x easier. Moreover I've changed slashes in cache key to dashes as this is what we use in 7.x stack.

## Justification

Using persistence cache here might be seen as a workaround, but I rather see it as a clean solution.
Persistence layer goes beyond just a database. In fact it is any storage, which includes filesystem, AWS, NFS/DFS, etc, therefore some information about files can be cached in this layer.

## Concerns about cache invalidation

@kmadejski in his solution ensured that cache key relies on Version Id and Field Id, so when an image field is updated with new image, different cache key will be used. For 6.x stack it might lead to some obsolete entries. When merged up to 7.x we can try to use tags to invalidate obsolete cache. This is the same case as for every cache key where Version Id / No and/or Field Id is used.

It is true that in case of Redis, any change to a file bypassing our Repository (Liip configuration change, manual change of a file) will require clearing persistence cache, however it's worth to mention that a solution involving storing this data in a database would yield the same result.

\* We need to document that manual change of image variation properties (image itself or liip configuration) requires clearing persistence cache. It makes a difference in case of Redis or other external server // cc @DominikaK 

**TODO**:
- [x] Fix loading images via URIs (should be loaded via filesystem paths).
- [x] Fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

adamwojs commented on eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php:
```diff
@@ -118,12 +127,17 @@ public function getVariation(Field $field, VersionInfo $versionInfo, $variationN
 
         try {
+            $resolvedPath = $this->ioResolver->resolve($originalPath, $variationName);
+            $image = $this->imagine->open($resolvedPath);
```
As we discussed on the slack: we need to use a path to the generated image instead of URL. Downloading image over the network have a significant impact on the performance.

alongosz replied:
> s we discussed on the slack: we need to use a path to the generated image instead of URL. Downloading image over the network have a significant impact on the performance.

True, I'll fix it today the way we agreed.

alongosz commented:
@adamwojs updated according to your recent suggestions.

Changes:
- I've added yet another decorator for AliasGenerator (0e15af1) with injected Imagine, separating responsibility:
  - It checks if ImageVariation has width and height, if not:
  - It loads local filesystem cache path instead of calling `Imagine::open` on URI and sets the dimensions.
  - It sets other overlooked and easily available at this point parameters (`fileSize`, `mimeType`)
- The result of that operation is still cached (persistence cache) to avoid opening the same image multiple times (if image changes by the means of Repository, the cache key changes also - so no problem).
- For `original` variation width and height are already known, so they're used directly (80d0edd).
- Imagine is no longer needed on the lowest layer of `AliasGenerator`.

andrerom commented:
> The result of that operation is still cached (persistence cache) to avoid opening the same image multiple times 

nitpick: "Persistence cache" => "Stash Cache" or "Image Variation Cache" to be specific , this is not SPI/Persistence and not Core/Persistence/Cache ;)

gggeek commented:
Interesting discussion on cache naming / split.
I agree that stash-cache is not a good name, as it relies on (accidental) implementation detail.
Sf cache is also not good, as I tend to associate it with cache of config and tpls, which is not the same domain as this one.
To me, image-variation-cache is part of content-cache from a logical pov, so it should be closely related to SPI/Persistence...

andrerom commented:
> Stash Cache is not a good name because I want to reuse the same cache layer in 7.x. Therefore I've chosen "Persistence". Is there a better word describing both Stash and Symfony cache?

> I agree that stash-cache is not a good name, as it relies on (accidental) implementation detail.

Reason why I specified Stash here on this PR, which is implementation for 6.x which happens to use Stash. Is that intention for 7.x _(master)_, form my side, was to not put it in cache but rather make it... persist it. So it does not need to load in the image to calculate size every-time cache has been cleared / expired / cold cache on new deploys, as this is particular expensive to do on for instance S3. Same same like legacy which stored it in attribute table, but different. But like legacy it would then also have to have some mechanism to clear the size, or allow image variations to be re-created _(also updating size info)_.

How, I'm not sure of. But when asked about my preference, ^this is what I expressed.

Feel free to rename, but usage of persistence is miss-leading here given our current SPI naming.

kmadejski commented on eZ/Bundle/EzPublishCoreBundle/Imagine/Variation/AliasGeneratorDecorator.php:
```diff
@@ -0,0 +1,148 @@
+
+        if (null === $variation->width || null === $variation->height) {
+            $variationBinaryFile = $this->getVariationBinaryFile($field->value->id, $variationName);
+            $absoluteVariationPath = realpath($this->webRootDir . $variationBinaryFile->uri);
```
I'm sorry, but this is not working with DFS.

Please, consider a change to something like this:
```
$variationBinaryFile = $this->getVariationBinaryFile($field->value->id, $variationName);
$image = $this->imagine->load($this->ioService->getFileContents($variationBinaryFile));
$dimensions = $image->getSize();

return new ImageVariation(
...
```
Thanks to that IOService will be responsible for finding and loading a proper file using right adapter.
Then, you can remove `$webRootDir` variable.

kmadejski commented on eZ/Bundle/EzPublishCoreBundle/Resources/config/image.yml:
```diff
@@ -81,6 +83,22 @@ services:
+            - '@ezpublish.fieldType.ezimage.io_service'
+            - '@ezpublish.config.resolver'
+            - '@liip_imagine'
+            - '%webroot_dir%'
```
You can remove it as mentioned above.

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Imagine/Variation/AliasGeneratorDecorator.php:
```diff
@@ -0,0 +1,148 @@
+ * Alias Generator Decorator which ensures (using Imagine if needed) that ImageVariation has proper
+ * dimensions.
+ */
+class AliasGeneratorDecorator implements VariationHandler
```
> How about renaming it to `AliasDimensionsDecorator`? IMHO it will better describe what this class is used for.

Hmm, then to me it would mean "something that decorates class returning alias dimensions". What we have here is another layer decorating AliasGenerator returning the same result. 
But I agree naming is not the best here.

How about `ImagineAwareAliasGenerator`? Imagine is hardcoded in the namespace anyway.

alongosz commented:
@adamwojs @kmadejski I've addressed the issues you've mentioned.
Ready for another round of reviews.
Note: Alias Generator Tests class was refactored to avoid repeating code, please take look at it as well.

@andrerom changed title to refer to Image Variation Cache as agreed. There's still one commit message to change, I'll do it after final code reviews.

lserwatka commented:
@alongosz please merge it up

alongosz commented:
Merged up to 6.13.
Work in progress on 7.x in #2349. Something doesn't work on demo, so I need to check if this is related.
7.x has new cache stack, so this was rather expected.

homarx commented:
Hello everybody,
concerning this pull request I recognized some problem.
I developed an responsive image slider with thumbnail navigation and lightbox on an eZ 6.13.4 installation. In my twig templates I have many `ez_image_alias` calls resulting in around 200 function calls for building the whole slider. I only need the URI of the image variation, not width or height or anything else.

When I request the page with the slider after clearing the cache via console it leads to an fatal error:

PHP Fatal error:  Uncaught ImagickException: cache resources exhausted `' @ error/cache.c/OpenPixelCache/3945 in /xyz/vendor/imagine/imagine/lib/Imagine/Imagick/Imagine.php:109\nStack trace:\n#0 /xyz/vendor/imagine/imagine/lib/Imagine/Imagick/Imagine.php(109): Imagick->readimageblob('\\xFF\\xD8\\xFF\\xE0\\x00\\x10JFIF\\x00\\x01\\x01\\x00\\x00...')\n#1 /xyz/vendor/ezsystems/ezpublish-kernel/eZ/Bundle/EzPublishCoreBundle/Imagine/Variation/ImagineAwareAliasGenerator.php(85): Imagine\\Imagick\\Imagine->load('\\xFF\\xD8\\xFF\\xE0\\x00\\x10JFIF\\x00\\x01\\x01\\x00\\x00...')\n#2 /xyz/vendor/ezsystems/ezpublish-kernel/eZ/Bundle/EzPublishCoreBundle/Imagine/Cache/AliasGeneratorDecorator.php(69): eZ\\Bundle\\EzPublishCoreBundle\\Imagine\\Variation\\ImagineAwareAliasGenerator->getVariation(Object(eZ\\Publish\\API\\Repository\\Values\\Content\\Field) in /xyz/mytemplate.html.twig on line xyz

The bottleneck is, that for every requested image variation the file has to be loaded 
 to get width and height and some other information (ImagineAwareAliasGenerator.php(85)) although I don't need more information than the URI of the variation.

As workaround I implemented a custom twig function which I called "ez_image_alias_uri". It does exactly the same as "ez_image_alias" except that it doesn't check if there is a variation width and height. So it doesn't load any variation binary file. In my case it is an imense improving of performance.

How about a further parameter for ez_image_alias function? This parameter could take one or a list of particular variation properties like "uri", "fileName", "imageId" or "mimeType". In class ImagineAwareAliasGenerator it could be checked if it is necessary to load the variation file for returning the asked properties or if it isn't necessary.
